larq/optimizers_v1.py

```diff
@@ -0,0 +1,200 @@
+        base_lr_params = [p for p in params if p.name not in self.multipliers]
+
+        updates = []
+        base_lr = self.optimizer.lr
+        for param in mult_lr_params:
+            self.optimizer.lr = base_lr * self.multipliers[param.name]
+            updates.extend(self.optimizer.get_updates(loss, [param]))
+
+        self.optimizer.lr = base_lr
+        updates.extend(self.optimizer.get_updates(loss, base_lr_params))
+
+        return updates
+
+    def __getattr__(self, name):
+        return getattr(self.optimizer, name)
+
+    def get_config(self):
+        return {
+            "optimizer": {
+                "class_name": self.optimizer.__class__.__name__,
+                "config": self.optimizer.get_config(),
+            },
+            "multipliers": self.multipliers,
+        }
+
+    @classmethod
+    def from_config(cls, config, custom_objects=None):
+        optimizer = tf.keras.optimizers.deserialize(
+            config.pop("optimizer"), custom_objects=custom_objects
+        )
+        return cls(optimizer, config["multipliers"])
+
+
+@utils.register_keras_custom_object
+class Bop(tf.keras.optimizers.Optimizer):
+    """Binary optimizer (Bop).
+
+    Bop is a latent-free optimizer for Binarized Neural Networks (BNNs).
+
+    !!! example
+        ```python
+        optimizer = lq.optimizers.Bop(fp_optimizer=tf.keras.optimizers.Adam(0.01))
+
+        ```
+
+    # Arguments
+    fp_optimizer: a `tf.keras.optimizers.Optimizer`.
+    threshold: determines to whether to flip each weight.
+    gamma: the adaptivity rate.
+    name: name of the optimizer.
+
+    # References
+    - [Latent Weights Do Not Exist: Rethinking Binarized Neural Network Optimization](https://arxiv.org/abs/1906.02107)
+    """
+
+    def __init__(self, fp_optimizer, threshold=1e-5, gamma=1e-2, name="Bop", **kwargs):
+        super().__init__(**kwargs)
+
+        if not isinstance(fp_optimizer, tf.keras.optimizers.Optimizer):
+            raise TypeError(
+                f"Expected tf.keras.optimizers.Optimizer, received {type(fp_optimizer)}."
+            )
+
+        with tf.keras.backend.name_scope(self.__class__.__name__):
+
+            self.fp_optimizer = fp_optimizer
+            self.threshold = threshold
+            self.gamma = gamma
+
+    def get_updates(self, loss, params):
+        grads = self.get_gradients(loss, params)
+        self.updates = []
+
+        ms = [
+            tf.keras.backend.zeros(
+                tf.keras.backend.int_shape(p), dtype=tf.keras.backend.dtype(p)
+            )
+            for p in params
+        ]
+
+        fp_params = []
+
+        for p, g, m in zip(params, grads, ms):
+            if self.is_binary(p):
+                m_t = (1 - self.gamma) * m + self.gamma * g
+
+                self.updates.append(tf.assign(m, m_t))
+                self.updates.append(
+                    tf.assign(
+                        p, lq.quantizers.sign(-p * tf.sign(p * m_t - self.threshold))
```

Comment:
@lgeiger I feel the `lq.quantizers.sign` here is a bit confusing, since our quantizers usually modify the backpropagation. How about moving our `sign` to `lq.math.sign` (if we expect to implement more low-precision specific math) or aliasing it to `lq.sign`?

Reply:
I agree 👍 

I think we should move it to `lq.math.sign`. This is then consistent with TensorFlow which will move a lot of their top level math functions to `tf.math` as well (see [here](https://github.com/tensorflow/community/blob/master/rfcs/20180827-api-names.md#appendix-1-additional-endpoints)).

In the future we can add a heaviside step function to binarize with `{0, 1}` and similar things.

@koenhelwegen Can you make a PR to move this to `lq.math`?